src/lib/png/png.c

```diff
@@ -0,0 +1,992 @@
+												!= sizeof(png_head->interlace_method)) {
+		printf("Error : can't read IHDR chunk.\n");
+		return -1;
+	}
+	if (png_head->interlace_method != 0 && png_head->interlace_method != 1) {
+		printf("Error : invalid interlace method\n");
+		return -1;
+	}
+
+	if (read(fd, &crc, sizeof(crc)) != sizeof(crc)) {
+		printf("Error : can't read crc of IHDR chunk.\n");
+		return -1;
+	}
+
+	return 0;
+}
+
+/* Now lib support main chunks and this fun is stub for work with other chunks*/
+static void read_chunk(int fd, uint8_t *chunk_name) {
+	return;
+}
+
+static int png_read(int fd, struct png *png_data) {
+	struct png_header png_head;
+	uint8_t chunk_length[4];
+	uint8_t chunk_name[4];
+	int pos = IHDR_END;
+
+	/**
+	 * IDAT_flag
+	 * 1 - IDAT was successfull read and decoded
+	 * 0 - IDAT not yet met
+	 */
+	int flag_IDAT = 0;
+
+	/* read IHDR chunk and put file info in struct png_header */
+	if (read(fd, chunk_length, sizeof(chunk_length)) != sizeof(chunk_length)) {
+		printf("Error : can't read IHDR chunk.\n");
+		return -1;
+	}
+
+	if (convert(chunk_length, sizeof(chunk_length)) != IHDR_SIZE) {
+		printf("Error : invalid length of IHDR data.\n");
+		return -1;
+	}
+
+	if (read(fd, chunk_name, sizeof(chunk_name)) != sizeof(chunk_name)) {
+		printf("Error : can't read IHDR chunk.\n");
+		return -1;
+	}
+
+	if (memcmp(chunk_name, "IHDR", sizeof(chunk_name)) != 0) {
+		printf("Error : invalid IHDR chunk.\n");
+		return -1;
+	}
+
+	if (read_IHDR(fd, &png_head) != 0) {
+		return -1;
+	}
+
+	png_data->width = png_head.width;
+	png_data->height = png_head.height;
+
+	switch (png_head.color_type) {
+		case 1:
+			printf("TODO : support of pallete images didn't add.\n");
+			return -1;
+		case 2:
+			png_data->color = TRCL;
+			break;
+
+		case 4:
+			png_data->color = TRCLA;
+			break;
+
+		case 6:
+			png_data->color = TRCLA;
+			break;
+
+		default:
+			printf("Error : invalid color type.\n");
+			return -1;
+	}
+	/* malloc mem for image */
+	png_data->image = (uint8_t*)malloc(png_head.width * png_head.height * 4 * sizeof(uint8_t));
+
+	/**
+	 * Reading chunks until IEND
+	 * its simple version and support just important chunks
+	 * fun read_chunk is a stub
+	 * TODO add reading of other chunks
+	 */
+	while(true) {
+		lseek(fd, pos, SEEK_SET);
+		read(fd, chunk_length, 4);
+		read(fd, chunk_name, 4);
+
+		if (memcmp(chunk_name, "IDAT", 4) == 0) {
+			/* read, decode IDAT and fill image buffer */
+			if (read_IDAT(fd, png_data->image, convert(chunk_length,sizeof(chunk_length)),
+																			&png_head) != 0) {
+				printf("Error while read and decode IDAT.\n");
+				free(png_data->image);
+				return -1;
+			}
+			flag_IDAT = 1;
+		} else if (memcmp(chunk_name, "IEND", 4) == 0) {
+			if (!flag_IDAT) {
+				printf("Error : there is no IDAT chunk in the file\n");
+				free(png_data->image);
+				return -1;
+			}
+			break;
+		} else if (memcmp(chunk_name, "bKGD", 4) == 0) {
```

Comment:
Is it possible to shorten this list of `memcpy` calls? Iterate some array of chunk names for example

Reply:
now fb_draw_image check size of image
add support for your test png

Reply:
squashed and edited commits